internal/pkg/dns/dns.go

```diff
@@ -158,6 +160,83 @@ func (h *Handler) SetProxy(prxs []*proxy.Proxy) bool {
 // Stop stops and clears dns proxy selector.
 func (h *Handler) Stop() { h.SetProxy(nil) }
 
+// NewNodeHandler creates a new NodeHandler.
+func NewNodeHandler(next plugin.Handler, hostMapper HostMapper, logger *zap.Logger) *NodeHandler {
+	return &NodeHandler{next: next, mapper: hostMapper, logger: logger}
+}
+
+// HostMapper is a name to node mapper.
+type HostMapper interface {
+	ResolveAddr(ctx context.Context, name string) (netip.Addr, bool)
+}
+
+// NodeHandler try to resolve dns request to a node. If required node is not found, it will move to the next handler.
+type NodeHandler struct {
+	next   plugin.Handler
+	mapper HostMapper
+	logger *zap.Logger
+
+	enabled atomic.Bool
+}
+
+// Name implements plugin.Handler.
+func (h *NodeHandler) Name() string {
+	return "NodeHandler"
+}
+
+// ServeDNS implements plugin.Handler.
+func (h *NodeHandler) ServeDNS(ctx context.Context, wrt dns.ResponseWriter, msg *dns.Msg) (int, error) {
+	if !h.enabled.Load() {
+		return h.next.ServeDNS(ctx, wrt, msg)
+	}
+
+	idx := slices.IndexFunc(msg.Question, func(q dns.Question) bool { return q.Qtype == dns.TypeA || q.Qtype == dns.TypeAAAA })
+	if idx == -1 {
+		return h.next.ServeDNS(ctx, wrt, msg)
+	}
+
+	req := request.Request{W: wrt, Req: msg}
+
+	// Check if the request is for a node.
+	result, ok := h.mapper.ResolveAddr(ctx, req.Name())
+	if !ok {
+		return h.next.ServeDNS(ctx, wrt, msg)
+	}
+
+	var rtype uint16
+
+	if result.Is4() {
+		rtype = dns.TypeA
```

Comment:
we should answer by query, not by result I guess

we should filter the result by v4/v6 based on query